airbyte-server/src/main/java/io/airbyte/server/ConnectorDefinitionSpecBackfiller.java

```diff
@@ -0,0 +1,263 @@
+      try {
+        if (sourceDef.getSpec() == null) {
+          // if a source definition is not being used and is not in the seed, don't bother to attempt to fetch
+          // a spec for it; just delete.
+          if (!seedSourceRepos.contains(sourceDef.getDockerRepository()) && !connectorReposInUse.contains(sourceDef.getDockerRepository())) {
+            LOGGER.info(
+                "migrateAllDefinitionsToContainSpec - Source Definition {} does not have a spec, is not in the seed, and is not currently used in a connection. Deleting...",
+                sourceDef.getName());
+            deleteSourceDefinition(database, sourceDef.getSourceDefinitionId());
+            continue;
+          }
+
+          LOGGER.info(
+              "migrateAllDefinitionsToContainSpec - Source Definition {} does not have a spec. Attempting to retrieve spec...",
+              sourceDef.getName());
+          final SynchronousResponse<ConnectorSpecification> getSpecJob = schedulerClient
+              .createGetSpecJob(imageName);
+          if (getSpecJob.isSuccess()) {
+            LOGGER.info(
+                "migrateAllDefinitionsToContainSpec - Spec for Source Definition {} was successfully retrieved. Writing to the db...",
+                sourceDef.getName());
+            final StandardSourceDefinition updatedDef = Jsons.clone(sourceDef).withSpec(getSpecJob.getOutput());
+            configRepository.writeStandardSourceDefinition(updatedDef);
+            LOGGER.info(
+                "migrateAllDefinitionsToContainSpec - Spec for Source Definition {} was successfully written to the db record.",
+                sourceDef.getName());
+          } else {
+            final LogRead logRead = jobConverter.getLogRead(getSpecJob.getMetadata().getLogPath());
+            LOGGER.info(
+                "migrateAllDefinitionsToContainSpec - Failed to retrieve spec for Source Definition {}. Logs: {}",
+                sourceDef.getName(),
+                logRead.toString());
+            throw new RuntimeException(String.format(
+                "Failed to retrieve spec for Source Definition %s. Logs: %s",
+                sourceDef.getName(),
+                logRead.toString()));
+          }
+        }
+      } catch (final Exception e) {
+        trackSpecBackfillFailure(trackingClient, configRepository, sourceDef.getDockerRepository(), sourceDef.getDockerImageTag(), e);
+        LOGGER.info("migrateAllDefinitionsToContainSpec - Failed to backfill spec into Source Definition {}. Error: {}", sourceDef.getName(), e);
+        if (configs.getVersion31ForceUpgrade()) {
+          LOGGER.info(
+              "migrateAllDefinitionsToContainSpec - Force upgrade set to true. Deleting Source Definition {} and any associated connections/syncs...",
+              sourceDef.getName());
+          deleteSourceDefinition(database, sourceDef.getSourceDefinitionId());
+        } else {
+          failedBackfillImages.add(imageName);
+        }
+      }
```

Comment:
is there a reason you prefer using exception handling for flow control here? instead of:
1. throw
2. catch
3. conditional
i think it would be a better practice in java to do
1. conditional
2. throw
3. catch

Reply:
actually on second thought, i don't know that this try/catch is useful. would it makes sense to just move all of this logic up? can we just do something roughly like this. what is the exception throwing getting us?

```
         } else {
            final LogRead logRead = jobConverter.getLogRead(getSpecJob.getMetadata().getLogPath());
            LOGGER.info(
                "migrateAllDefinitionsToContainSpec - Failed to retrieve spec for Source Definition {}. Logs: {}",
                sourceDef.getName(),
                logRead.toString());
            trackSpecBackfillFailure(trackingClient, configRepository, sourceDef.getDockerRepository(), sourceDef.getDockerImageTag());
            if (configs.getVersion31ForceUpgrade()) {
              LOGGER.info(
                  "migrateAllDefinitionsToContainSpec - Force upgrade set to true. Deleting Source Definition {} and any associated connections/syncs...",
                  sourceDef.getName());
              deleteSourceDefinition(database, sourceDef.getSourceDefinitionId());
            } else {
              failedBackfillImages.add(imageName);
            }
          }
        }
```

Reply:
The reason why I structured it the way I did was because I thought that we needed to have a try-catch around the `schedulerClient.createGetSpecJob` call anyway, in case any errors were thrown in the job, so that we would catch those errors and then execute the following logic accordingly.

However looking back on this, I realize that if there is an error in the getSpec job, it should not manifest as a thrown exception, but will rather be indicated by `getSpecJob.isSuccess()` being `false`, as I have written the rest of the code to reflect. So I think what you are suggesting makes sense, and we shouldn't need the try catch. I will refactor

Reply:
I have pushed the refactor and I think it looks a lot better now! Thanks for the suggestion